### PR TITLE
[ffwizard] never try to auto-login if password has been set

### PR DESCRIPTION
**Pull request by sarumpaet**
If a user sets a password but doesn't finish the set up process in
the wizard, the initial ffwizard auto-login redirect fails. This
adds a flag so that no auto-login happens if a password has been set.
Fixes https://github.com/freifunk-berlin/firmware/issues/208

### CODE REVIEW

**Comment by lynxis** [comment]
LGTM but I don't usually code lua. merge if you like.

**Comment by SvenRoederer** [comment]
Why not checking if the password is set in /etc/password (/etc/shadow)?
Having this in flag in UCI ffwizard will not work if the user changed the password via console or regular web-UI.

**Comment by sarumpaet** [comment]
Look at the comment in index.htm: The view isn't run as root and does not have access to /etc/shadow.
The approach does not cover all the possible cases, but it DOES cover all the REPORTED cases I think, plus it doesn't require another ugly OpenWrt patch. I'd say if someone wants to fix it upstream, go ahead, until then we should just merge this, it'll be good enough.

**Comment by SvenRoederer** [comment]
@sarumpaet : btw. when not running as root, as which user the script is running ? Just for my curiosity

**Comment by SvenRoederer** [comment]
what about ubus + rpcd-plugin:
```
root@sam0815-test:~# cat /usr/libexec/rpcd/freifunk_berlin 
#!/bin/sh

# taken from https://wiki.openwrt.org/doc/techref/rpcd
case "$1" in
        list)
                echo '{ "root_has_pwd": {} }'
        ;;
        call)
                case "$2" in
                        root_has_pwd)
                                local USER=root
                                # taken from http://serverfault.com/a/240961
                                grep ${USER} /etc/shadow | awk -F":" '{ if ($2 == "!" || $2 == "*") \
                                                print "\{\"password_is_set\": \"no\"\}"; \
                                        else \
                                                print "\{\"password_is_set\": \"yes\"\}"\
                                        }'
                        ;;
                esac
        ;;
esac
```
This can creates an object in ubus which can be queried:
> root@sam0815-test:~# ubus call -S -v freifunk_berlin root_has_pwd 
{"password_is_set":"yes"}

And then via libubus to lua:
```
 root@sam0815-test:~# cat /root/check_for_rootpwd.lua 
require "ubus"

local conn = ubus.connect()
if not conn then
    error("Failed to connect to ubusd")
end

local status = conn:call("freifunk_berlin", "root_has_pwd", {})
for k, v in pairs(status) do
    print("key=" .. k .. " value=" .. tostring(v))
end

conn:close()
```

gives 
>root@sam0815-test:~# lua check_for_rootpwd.lua 
key=password_is_set value=yes


This would work no matter which way the password was set.

**Comment by sarumpaet** [comment]
Wow. This adds two scripts and quite a bit of complexity and brittleness (how long until some OpenWrt update will break this?) to cover a case that doesn't matter in practice and could be easily fixed with about one line in upstream (see the very first comment in https://github.com/freifunk-berlin/firmware/issues/208 ). I applaud the effort, but how about directing this energy into something like https://github.com/freifunk-berlin/firmware/issues/292 ? :)

**Comment by SvenRoederer** [comment]
only one plugin for rpcd and a change of the code in the ffwizard, that will be changed anyway.

I don't think that the rpcd will be canceled in the near future, so not expecting it to becom incompatible.

**Comment by booo** [comment]
I favor a solution that fixes the problem once and for all. Either we try to implement the proposal from https://github.com/freifunk-berlin/firmware/issues/208 or we take the rpcd solution.

Anyway I think we should solve this in the next release. It's anyway broken since a few releases and we should really push a new release now!

**Comment by SvenRoederer** [comment]
+1 for keeping this for the next release

**Comment by bobster-galore** [comment]
I think we should decide regarding our resources if this is easily finished in one way or another we should include it in 0.2.0. If not we can postpone it to a later release. cite "to cover a case that doesn't matter in practice"
To cut it short if there is no surplus of bored willing programmers we should keep it in mind for the next release.

**Comment by sarumpaet** [comment]
I'd just merge the PR as-is, then replace it by a "complete" solution later if need be.

**Comment by lynxis** [comment]
This is just a small fix, I'm going to merge it.
@SvenRoederer if you want to do it that way, it would also introduce a dependency rpcd which is not yet part of firmware. But might be in the future.

**Comment by SvenRoederer** [comment]
@lynxis There is no direct dependency on rpcd, but luci-base has one. So it's included already anyways.

**Comment by lynxis** [comment]
@SvenRoederer ok, I see. I only looked into our old release, which did not contain rpcd. BTW. The script also contain a bug when the password string is empty.

This PR is just a quick and simple fix for a problem which will be solved by another wizard. I don't see big problems with this PR.

**Comment by SvenRoederer** [comment]
https://github.com/freifunk-berlin/firmware-packages/pull/83#issuecomment-232204595 also fail on "LuCI not running as root", as the code can not access the UBus-socket, which is mode 600 for user ang group "root".

Going this way will require to go via "uhttp-mod-ubus" and implementing some ACL on UBus/RPC for this.

**Comment by sarumpaet** [comment]
Let's merge this? Just tested again, works.

In case the user set the password via SSH or something before running the wizard, he can still log in just by deleting the URL parameters (i.e., by going to `/cgi-bin/luci/admin/freifunk/assistent`) or by clicking "Administration" to the bottom right.

**Comment by SvenRoederer** [comment]
do we need a migration for this (uci:set("ffwizard", "settings", "passwordset", "true")) ?

if so maybe inverting the logic might be an option. 
to change to option to "passwordunset" and having it "true" after initial flash, so current setups are not affected automatically

**Comment by sarumpaet** [comment]
No migration is necessary. For routers with Kathleen 0.1.2 config, runbefore is set to "true", so both IFs will correctly bail out on the "if not runbefore..." already.

File: utils/luci-mod-freifunk-ui/luasrc/view/freifunk/index.htm [hunk 1]
```diff
@@ -16,12 +16,13 @@
 local ipkg = require "luci.model.ipkg"
 
 -- only redirect if assistent is installed and no root password is set
--- we use isfile because the view is run as root and we can't use e.g.
--- ipkg.installed or checkpasswd
+-- we use isfile because the view is not run as root and we can't use
```
Comment (SvenRoederer) [inline utils/luci-mod-freifunk-ui/luasrc/view/freifunk/index.htm]: can you put this in a separate commit to get this more easy merged to upstream?

Comment (sarumpaet) [inline utils/luci-mod-freifunk-ui/luasrc/view/freifunk/index.htm]: There is no upstream for this; the "redirect to assistent" thing is Berlin specific.

Comment (SvenRoederer) [inline utils/luci-mod-freifunk-ui/luasrc/view/freifunk/index.htm]: i should have checked this before ... did not believe that we are so much apart from upstream

**Comment by booo** [comment]
@sarumpaet Can you review the latest changes and merge this if you are happy with them?

To me the final solution looks good.

**Comment by sarumpaet** [comment]
Fixed in https://github.com/freifunk-berlin/firmware-packages/pull/89
